### ISSUE TITLE
OCSP responder certificates are not actually linted, to date (fix needed)

### ISSUE THREAD

**Issue by defacto64** [zmap/zlint#838]
I think I found a problem in an already existing lint, and a more general problem in a core component of Zlint. 

Let me start from the more general problem. I think we all agree that it is (would be) a good thing to be able to also lint OCSP responder certificates, since there are specific requirements for them in the CABF baseline requirements (all of them). For a start, an OCSP responder cert must contain the ocsp-nocheck extension, so I was happy to see that there already exists a lint [1] that checks for this. However, I did some tests with real OCSP responder certs, both good and bad, and I found with dismay that that lint is actually never invoked (and therefore it's as if it didn't exist). After a bit of digging, I understood that that no OCSP responder cert is really linted by Zlint, to date, for the simple reason that a lint that refers to the BRs is considered not applicable (NA) if the certificate to be linted is not a ServerAuthCert. This is the consequence of this code excerpt from **lint/base.go**:

```
func (l *CertificateLint) Execute(cert *x509.Certificate, config Configuration) *LintResult {
	if l.Source == CABFBaselineRequirements && !util.IsServerAuthCert(cert) {
		return &LintResult{Status: NA}
	}

```
The problem stems from the fact that, normally, an OCSP responder cert does NOT contain the serverAuth EKU (even less the AnyKeyUsage EKU), which is the test done by `util.IsServerAuthCert()`. And besides, that is _forbidden_ since CABF BR 2.0.0.

The same check is re-done (uselessly) in the lint [1], and would cause the lint to be skipped regardless of the above issue in lint/base.go.

I think this problem should be fixed soon. 

Does anyone have any suggestions? Otherwise I can see if I can come up with something....

_______________________________
[1] https://github.com/zmap/zlint/blob/master/v3/lints/cabf_br/lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included_server_auth.go

**Comment by christopher-henderson** [comment]
Good catch @defacto64 

Right, so this is tricky for quite a few reasons.

1. CABF/BR, in general, operates on server certificates so we certainly do not want to remove the check from the framework as this would imply that every individual lint would be responsible for performing this check themselves.
2. The requirement is, itself, quite awkward. It is a requirement on an OCSP certificate in a sea of server certificate requirements.
3. Indeed, if I do a case insensitive grep for `ocsp` I find [lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included_server_auth.go](https://github.com/zmap/zlint/blob/master/v3/lints/cabf_br/lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included_server_auth.go) and possibly [lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included_server_auth.go](https://github.com/zmap/zlint/blob/master/v3/lints/cabf_br/lint_ca_digital_signature_not_set.go).

We have hundreds of lints for server certificates and one-to-two (that I saw) lints for OCSP certificates.

I tried a [quick-and-dirty hack](https://github.com/zmap/zlint/pull/842/files) that says `okay, fine, they all have to be server certificates EXCEPT this one lint`.

```go
if l.Source == CABFBaselineRequirements {
    _, ocspCertLint := l.Lint().(*cabf_br.OCSPIDPKIXOCSPNocheckExtNotIncludedServerAuth)
    if !util.IsServerAuthCert(cert) && !ocspCertLint {
        return &LintResult{Status: NA}
    }
}
```

This, however, is an import cycle as the `cabf_br` package refers to the `lint` package for items such as `lint.RegisterLint`. `lint.Metadata`, `lint.Result`, and so on.

---
Perhaps I will have to think on it further, but off the top of my head in order to get this lint operational our options include.

* Create a sub-category (`cabf_br::ocsp`).
* Create a new (optional) interface wherein an individual lint could choose to override the framework's `effective` check.

I will say that I (personally) have no appetite for #1 as it creates a whole new category for, effectively, a single lint.

#2 is somewhat more appetizing as I can see it being used by other lints with similar awkward nuances. Something perhaps akin to...

```go
type Overrider interface {
	LintInterface
	OverrideFrameworkCheck(c *x509.Certificate) *LintResult
}
...
...
...
override, ok := l.Lint().(Overrider)
if ok {
	result := override.OverrideFrameworkCheck(cert)
	if result != nil {
		return result
	}
} else {
	if l.Source == CABFBaselineRequirements && !util.IsServerAuthCert(cert) {
		return &LintResult{Status: NA}
	}
	if l.Source == CABFSMIMEBaselineRequirements && !((util.IsEmailProtectionCert(cert) && util.HasEmailSAN(cert)) || util.IsSMIMEBRCertificate(cert)) {
		return &LintResult{Status: NA}
	}
}
```

I've hacked up something (with terrible names and little consideration for implications) at [#842](https://github.com/zmap/zlint/pull/842/files).

**Comment by defacto64** [comment]
I think I see how the thing works. Please correct me if I am wrong.
If a linter implements the `Overrider `interface, the framework check is skipped.
However, in this case I understand that the linter can perform its job either within the `OverrideFrameworkCheck` function itself (by returning a `LintResult`) or return `nil `and then do its job within the usual `Execute `function. Right?
If so, what's the use of this double option? Wouldnt it suffice to just check if the linter implements the `Overrider `interface, skip the the framework check in case, and let the linter do its job via the `Execute `function alone, as usual?

**Comment by christopher-henderson** [comment]
That is certainly an option! I admit that I didn't put _too_ much thought into the details just yet, just spit balling on ideas that could accomplish the goal without having to tear up too much just for one (or several) lints.

```go
type AGoodAndDescriptiveName interface {
    func Override()
}

...
...
...
if _, ok := lint.(AGoodAndDescriptiveName); !ok {
    // framework check
}
lint.Execute(cert)

```

**Comment by toddgaunt-gs** [comment]
Hey @christopher-henderson, I'm curious what other cases could be solved by this interface check. I'm wondering if a bool or other flag value in the `CertificateLint` struct could accomplish the same thing for this one lint by simply returning nil. Something to consider I think unless there is good reason as you've metnioned in a previous comment. The main reason is that I am not seeing that this interface accomplishes anything that having a flag to bypass the NA check and modifying `CheckApplies` can't accomplish. Having an additional interface and optional method just seems more complicated than its worth to me. Let me know if I am missing anything here, since otherwise it seems like your solution will do the trick.

**Comment by christopher-henderson** [comment]
@toddgaunt-gs adding a default-false field to [CertificateLint](https://github.com/zmap/zlint/blob/456dc01dad591ddaaf005f6a955fbca032379c0f/v3/lint/base.go#L175) is indeed another option.

**Comment by defacto64** [comment]
Yes, a default-false booean in **`LintMetadata`** seems simpler. Could be like follows:

`SkipFrameworkCheck bool`

Then, the framework check in **base.go** could be modified like this:
```

func (l *CertificateLint) Execute(cert *x509.Certificate, config Configuration) *LintResult {
    if !l.SkipFrameworkCheck {
        if l.Source == CABFBaselineRequirements && !util.IsServerAuthCert(cert) {
            return &LintResult{Status: NA}
        }
        if l.Source == CABFSMIMEBaselineRequirements && !((util.IsEmailProtectionCert(cert) && util.HasEmailSAN(cert)) || util.IsSMIMEBRCertificate(cert)) {
            return &LintResult{Status: NA}
        }
    }
    lint := l.Lint()
    err := config.MaybeConfigure(lint, l.Name)
    if err != nil {
	return &LintResult{
		Status:  Fatal,
		Details: err.Error()}
    }
    ...and so on...
```

How about that?

**Comment by christopher-henderson** [comment]
Howdy @bineeshav.

Thank you for your patience. I admit that I am often quite pressed for time-and-energy and that this ticket remained in limbo die to analysis-by-paralysis in #842.

I have reviewed the issue and I remain convinced of the implementation as proposed due to its simplicity and limited scope. I am inclined to move forward with the solution ideated with @defacto64. I will, however, allow @defacto64 an opportunity to also chime in (if he so pleases) as he opened the original issue.

**Comment by christopher-henderson** [comment]
Thank you for your patience @bineeshav and @defacto64. I have merged #842
